\documentclass{bioinfo}
\copyrightyear{2015} \pubyear{2015}


%https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3463124/

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Manuscript Category}

\begin{document}
\firstpage{1}

\subtitle{Subject Section}

\title[Rethomics]{Rethomics: an R framework to analyse high-throughput behavioural data}
\author[Geissmann \textit{et~al}.]{Quentin Geissmann\,$^{\text{\sfb 1,}*}$, Luis Garcia\,$^{\text{\sfb 2}}$, Esteban Beckwith\,$^{\text{\sfb 1}}$,  and Giorgio Gilestro\,$^{\text{\sfb 1,}*}$}
\address{$^{\text{\sf 1}}$Department, Institution, City, Post Code, Country and \\
$^{\text{\sf 2}}$Department, Institution, City, Post Code,
Country.}

\corresp{$^\ast$To whom correspondence should be addressed.}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\abstract{
	\textbf{Motivation:}
		Ethomics, a quantitative and high-throughput approach to ethology, is a novel and exciting field.
		The recent development of methods that automatically score variables in multiple animals provides an unprecedented insight into the study of behaviours. 
		The analysis of ethomics data presents many challenges that are conceptually independent of the acquisition platform.
		However, there is, little effort in providing a general framework to analyse multiple and long behavioural time series.
		We developped the \texttt{rethomics} framework, a suite of R packages that altogether provide utilities to:
		import, store, visualise and analyse behavioural data.\\
	\textbf{Availability and Implementation:}
	All packages in the \texttt{rethomics} framework are available under the terms of the GPLv3 license.
	Exaustive installation instructions and tutorials are available at
 		\href{https://github.com/rethomics}{https://github.com/rethomics/} 
 		\href{https://rethomics.github.io/}{https://rethomics.github.io/}.\\
	\textbf{Contact:}
		\href{qgeissmann@gmail.com}{qgeissmann@gmail.com}
	}

\maketitle

\section{Introduction}
Animal behaviours are complex phenotypes that are manifestations of the interaction between the nervous system and the external or internal environment.
In the last few decades, our ability to record large amounts of various phenotypical data has tremendously increased.
Behaviour scoring is certainly not an exception to this trend.
Indeed, many platforms (todo citations) have been developed in order to allow biologists to continuously record behaviours such as activity, position and feeding of multiple animals over long durations (days or weeks).

The availability of ever so large amounts of data is very exciting as it paves the way for refined analyses.
Clearly, the multiplicity of model organisms, hypotheses and paradigms makes the diversity of recording tools crucial.
However, regarding subsequent data analysis, there is no unified, programmatic, framework that could be used as a set of building blocks in a pipeline.
Instead, tools tend to consist of graphical interfaces with strictly defined functionalities that only import data from a single platform.
There are three issues with this approach.
First of all, state of the art analysis and visualisation requires a flexibility than only a programmatic interface can provide.
Secondly, it favours replicated work as developers need to provide their own implementation to very similar problems.
Lastly, it links analysis and visualisation to the target platform, which makes it very difficult to share cross-platform tools.
 
Thankfully, behavioural data is conceptually largely agnostic of the acquisition platform and paradigm. 	
Typically, the behaviour of each individual is a long time series (possibly multivariate and heterogeneous).
In addition, each individual has to be unambiguously identified and associated with arbitrary metadata defined by the experimenter (sex, treatment, genotype and so on). Efficiently combining and manipulating these information, on datasets of hundreds of individuals, each recorded for weeks, is not trivial. The availability of such a tool help promoting the analysis of behaviour as a data science.

Here, we describe \texttt{rethomics}, a framework that unifies analysis of large behavioural dataset in an efficient and manner.
It offers an elegant solution to store, manipulate and visualise a large amount of behavioural data.
Rethomics comes with a extensive documentation and a set of both practical and theoretical tutorials.



%\enlargethispage{12pt}

\section{Approach}
%Figure~2\vphantom{\ref{fig:02}} shows that the above method  Text
\texttt{rethomics} is implemented as a collection of small packages linked to one another.
This development model follows modern frameworks such as the \texttt{tidyverse}, which results in increases testability and maintainability.
The different tasks of the analysis workflow (\emph{i.e.} data import, manipulation and visualisation)
are explicitly handled by different packages (Fig.~1\vphantom{\ref{fig:01}}).
At the core of \texttt{rethomics}, the \texttt{behavr} package offers an very flexible and efficient solution to store both large amounts data (\emph{e.g.} position and activity) and metadata (\emph{e.g.} treatment, genotype and so on) in a single \texttt{data.table} derived object.
Any input package will import experimental data as a \texttt{behavr} table which can, in turn, be manipulated and visualised regardless of the input platform.
Analyses results and plots integrate seamlessly within the \texttt{R} ecosystem, hence providing users with state visualisation and statistics tools.


\section{Framework  features}
\subsection{Data import}
xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl 
\subsection{Internal data structure}
xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl 
\subsection{Circadian analysis}
xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl 
\subsection{Sleep scoring}
xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl 
\subsection{Visualisation}
xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl 


\section{Conclusion}

xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl 
xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl 
xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl xxx xxxxx xxxx dwjdkl 

\section*{Acknowledgements}

Han Kim. Patrick Kr{\"a}tschmer

\section*{Funding}

This work has been supported by the... Text Text  Text Text.\vspace*{-12pt}

%\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
%\bibliography{Document}


\begin{thebibliography}{}

\bibitem[Bofelli {\it et~al}., 2000]{Boffelli03}
Bofelli,F., Name2, Name3 (2003) Article title, {\it Journal Name}, {\bf 199}, 133-154.

\bibitem[Bag {\it et~al}., 2001]{Bag01}
Bag,M., Name2, Name3 (2001) Article title, {\it Journal Name}, {\bf 99}, 33-54.

\bibitem[Yoo \textit{et~al}., 2003]{Yoo03}
Yoo,M.S. \textit{et~al}. (2003) Oxidative stress regulated genes
in nigral dopaminergic neurnol cell: correlation with the known
pathology in Parkinson's disease. \textit{Brain Res. Mol. Brain
Res.}, \textbf{110}(Suppl. 1), 76--84.

\bibitem[Lehmann, 1986]{Leh86}
Lehmann,E.L. (1986) Chapter title. \textit{Book Title}. Vol.~1, 2nd edn. Springer-Verlag, New York.

\bibitem[Crenshaw and Jones, 2003]{Cre03}
Crenshaw, B.,III, and Jones, W.B.,Jr (2003) The future of clinical
cancer management: one tumor, one chip. \textit{Bioinformatics},
doi:10.1093/bioinformatics/btn000.

\bibitem[Auhtor \textit{et~al}. (2000)]{Aut00}
Auhtor,A.B. \textit{et~al}. (2000) Chapter title. In Smith, A.C.
(ed.), \textit{Book Title}, 2nd edn. Publisher, Location, Vol. 1, pp.
???--???.

\bibitem[Bardet, 1920]{Bar20}
Bardet, G. (1920) Sur un syndrome d'obesite infantile avec
polydactylie et retinite pigmentaire (contribution a l'etude des
formes cliniques de l'obesite hypophysaire). PhD Thesis, name of
institution, Paris, France.

\end{thebibliography}
\end{document}
